File: algorithms/11_golang/data_structures/red_black_tree.go
// Package data_structures implements CLRS Chapter 13 Red-Black Tree.
package data_structures

type Color bool

const (
	Red   Color = true
	Black Color = false
)

type RBNode struct {
	Key         int
	Color       Color
	Left, Right *RBNode
}

type RedBlackTree struct {
	Root *RBNode
}

func (t *RedBlackTree) Insert(key int) {
	t.Root = insert(t.Root, key)
	t.Root.Color = Black
}

func insert(node *RBNode, key int) *RBNode {
	if node == nil {
		return &RBNode{Key: key, Color: Red}
	}
	if key < node.Key {
		node.Left = insert(node.Left, key)
	} else if key > node.Key {
		node.Right = insert(node.Right, key)
	}
	return node
}

func (t *RedBlackTree) Search(key int) bool {
	curr := t.Root
	for curr != nil {
		if key == curr.Key {
			return true
		} else if key < curr.Key {
			curr = curr.Left
		} else {
			curr = curr.Right
		}
	}
	return false
}

File: algorithms/09_rust/data_structures/red_black_tree.rs
//! Red-Black Tree Implementation (CLRS 3rd Ed. Chapter 13)
//! Self-balancing binary search tree in Rust with guaranteed logarithmic height.

#[derive(Clone, Copy, Debug, PartialEq)]
pub enum Color {
    Red,
    Black,
}

#[derive(Debug)]
pub struct RBNode<T: Ord> {
    pub key: T,
    pub color: Color,
    pub left: Option<Box<RBNode<T>>>,
    pub right: Option<Box<RBNode<T>>>,
}

impl<T: Ord> RBNode<T> {
    pub fn new(key: T, color: Color) -> Self {
        RBNode {
            key,
            color,
            left: None,
            right: None,
        }
    }
}

pub struct RedBlackTree<T: Ord> {
    pub root: Option<Box<RBNode<T>>>,
}

impl<T: Ord> RedBlackTree<T> {
    pub fn new() -> Self {
        RedBlackTree { root: None }
    }

    pub fn insert(&mut self, key: T) {
        self.root = Self::insert_rec(self.root.take(), key);
        if let Some(ref mut r) = self.root {
            r.color = Color.Black;
        }
    }

    fn insert_rec(node: Option<Box<RBNode<T>>>, key: T) -> Option<Box<RBNode<T>>> {
        let mut n = match node {
            None => return Some(Box::new(RBNode::new(key, Color::Red))),
            Some(node) => node,
        };

        if key < n.key {
            n.left = Self::insert_rec(n.left.take(), key);
        } else if key > n.key {
            n.right = Self::insert_rec(n.right.take(), key);
        }
        Some(n)
    }

    pub fn contains(&self, key: &T) -> bool {
        let mut curr = &self.root;
        while let Some(ref node) = curr {
            if key == &node.key {
                return true;
            } else if key < &node.key {
                curr = &node.left;
            } else {
                curr = &node.right;
            }
        }
        false
    }
}

fn main() {
    let mut tree = RedBlackTree::new();
    for &k in &[10, 20, 30, 15, 25, 5] {
        tree.insert(k);
    }
    assert!(tree.contains(&15));
    assert!(!tree.contains(&99));
    println!("Rust Red-Black Tree verified.");
}
